coalib/core/Core.py

```diff
@@ -0,0 +1,175 @@
+    :param event_loop:
+        The asyncio event loop to schedule bear tasks on.
+    :param running_tasks:
+        Tasks that are already scheduled, organized in a dict with
+        bear instances as keys and asyncio-coroutines as values containing
+        their scheduled tasks.
+    :param executor:
+        The executor to which the bear tasks are scheduled.
+    """
+    for bear in bears:
+        tasks = {
+            event_loop.run_in_executor(
+                executor, bear.execute_task, bear_args, bear_kwargs)
+            for bear_args, bear_kwargs in bear.generate_tasks()}
+
+        running_tasks[bear] = tasks
+
+        for task in tasks:
+            task.add_done_callback(functools.partial(
+                finish_task, bear, result_callback,
+                running_tasks, event_loop, executor))
+
+        logging.debug('Scheduled {!r} (tasks: {})'.format(bear,
+                                                          len(tasks)))
+
+        if not tasks:
+            # We need to recheck our runtime if something is left to
+            # process, as when no tasks were offloaded the event-loop could
+            # hang up otherwise.
+            cleanup_bear(bear, result_callback, running_tasks, event_loop,
+                         executor)
+
+
+def finish_task(bear,
+                result_callback,
+                running_tasks,
+                event_loop,
+                executor,
+                task):
+    """
+    The callback for when a task of a bear completes. It is responsible for
+    checking if the bear completed its execution and the handling of the
+    result generated by the task.
+
+    :param bear:
+        The bear that the task belongs to.
+    :param result_callback:
+        A callback function which is called when results are available.
+    :param running_tasks:
+        Dictionary that keeps track of the remaining tasks of each bear.
+    :param event_loop:
+        The ``asyncio`` event loop bear-tasks are scheduled on.
+    :param executor:
+        The executor to which the bear tasks are scheduled.
+    :param task:
+        The task that completed.
+    """
+    try:
+        results = task.result()
+    except Exception as ex:
+        # FIXME Try to display only the relevant traceback of the bear if error
+        # FIXME occurred there, not the complete event-loop traceback.
+        logging.error('An exception was thrown during bear execution.',
+                      exc_info=ex)
+
+        results = None
+    finally:
+        running_tasks[bear].remove(task)
+        cleanup_bear(bear, result_callback, running_tasks, event_loop,
+                     executor)
+
+    if results is not None:
+        for result in results:
+            try:
+                # FIXME Long operations on the result-callback could block the
+                # FIXME   scheduler significantly. It should be possible to
+                # FIXME   schedule new Python Threads on the given event_loop
```

Comment:
Actually I'm not sure how result-callbacks are handled, but it could be that they are executed sequentially (which I currently assume). Offloading them into Python threads makes them at least executing "in parallel" (not real-time, but Python switches those threads regularly).